### PR TITLE
chore(plugins-view): plugin list was not displayed for DevWorkspace instances

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

### What does this PR do?
Implements [`K8sWorkspaceServiceImpl.getWorkspaceSettings()` method](https://github.com/eclipse-che/che-theia/blob/main/extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts#L74) to be able Che-Theia get plugin registry internal/external URLs.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/20448

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)
- [ ] Optional Companion PR for updating the HappyPath tests is approved and ready to be merged

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

File: extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts [hunk 1]
```diff
@@ -71,8 +71,12 @@ export class K8sWorkspaceServiceImpl implements WorkspaceService {
   }
 
   public async getWorkspaceSettings(): Promise<WorkspaceSettings> {
-    console.log('workspaceService.getWorkspaceSettings() not supported');
-    return {};
+    console.log('workspaceService.getWorkspaceSettings() is implemented partially');
+
+    return {
+      CHE_PLUGIN_REGISTRY_URL: this.env.getPluginRegistryURL(),
```
Comment (benoitf) [inline extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:77]: note that these variables are available in Che/DevWorkspaceOperator mode but not in pure DevWorkspace Operator mode so please check that there is no error on pure DevWorkspace

Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:77]: then, how can we get links to plugin registry in pure DevWorkspace Operator mode?

Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:77]: config map does not contain internal UR

Screenshot from 2021-10-05 18-13-21

Comment (benoitf) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:77]: we could display the notification or use online instance of the registry at github.io

File: extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-devworkspace-env-variables.ts [hunk 2]
```diff
@@ -40,32 +40,58 @@ export class K8sDevWorkspaceEnvVariables {
...
     }
+
     if (process.env.PROJECTS_ROOT === undefined) {
       console.error('Environment variable PROJECTS_ROOT is not set');
     } else {
       this.projectsRoot = process.env.PROJECTS_ROOT;
     }
+
+    if (process.env.CHE_PLUGIN_REGISTRY_URL === undefined) {
+      console.error('Environment variable CHE_PLUGIN_REGISTRY is not set');
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-devworkspace-env-variables.ts:85]: CHE_PLUGIN_REGISTRY_**URL**

File: extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts [hunk 3]
```diff
@@ -71,8 +71,17 @@ export class K8sWorkspaceServiceImpl implements WorkspaceService {
...
   public async getWorkspaceSettings(): Promise<WorkspaceSettings> {
-    console.log('workspaceService.getWorkspaceSettings() not supported');
-    return {};
+    const publicURI = this.env.getPluginRegistryURL();
+    const privateURI = this.env.getPluginRegistryInternalURL() || publicURI;
+
+    if (publicURI && privateURI) {
+      return {
+        CHE_PLUGIN_REGISTRY_URL: publicURI,
+        CHE_PLUGIN_REGISTRY_INTERNAL_URL: privateURI,
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:80]: If we call these keys without `CHE_` prefix, as `PLUGIN_REGISTRY_URL` and `PLUGIN_REGISTRY_INTERNAL_URL`, then we don't need any changes in `che-plugin-service.ts`. Isn't it?

File: extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts [hunk 4]
```diff
@@ -71,8 +71,17 @@ export class K8sWorkspaceServiceImpl implements WorkspaceService {
   }
 
   public async getWorkspaceSettings(): Promise<WorkspaceSettings> {
-    console.log('workspaceService.getWorkspaceSettings() not supported');
-    return {};
+    const publicURI = this.env.getPluginRegistryURL();
+    const privateURI = this.env.getPluginRegistryInternalURL() || publicURI;
```
Comment (azatsarynnyy) [inline extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:75]: Is there a real reason for using a public URI as a private one if it is not set?
Considering we're already checking that in https://github.com/eclipse-che/che-theia/blob/f677d115a50ab18cc4ae3ac25ce7236759d2f907/extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts#L103

File: extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts (hunk 3 above)
Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:80]: > PLUGIN_REGISTRY_URL in che-plugin-service.ts means a bit different

Actually, it means the same - URL of the plugin registry. It doesn't matter how this registry is implemented and where it's located, etc.

`workspaceService.getWorkspaceSettings()` allows to abstract from where it's stored: either Che-Server WS settings or DevWorkspace env. variables.

So, if we get rid of the `CHE_` prefixes, we can make it more transparent in `che-plugin-service.ts`.

Comment (benoitf) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:80]: looks like the returned object has the same key but with different values ?

Comment (vitaliy-guliy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts]: > looks like the returned object has the same key but with different values ?

I would say different keys with the same meaning.

File: extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts (hunk 4 above)
Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:75]: I'm not asking about both URIs.
I'm asking about both places of using it:
1. https://github.com/eclipse-che/che-theia/blob/4f6995d3ab8aad15669845e2de5b1464f79f8933/extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts#L75
2. https://github.com/eclipse-che/che-theia/blob/f677d115a50ab18cc4ae3ac25ce7236759d2f907/extensions/eclipse-che-theia-plugin-ext/src/node/che-plugin-service.ts#L103

In other words, if I want to get and see the "raw" settings returned by `workspaceService.getWorkspaceSettings()`, I should be able to get it as it is. Without any substitutions inside, like in 1.
And only when I get the "raw" settings, I can do with them whatever I want - like making such checks as in 2.

Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:75]: > Private URI is used to get the plugin list on the backend, public URI is used by frontend to load plugin icons

`workspaceService.getWorkspaceSettings()` knows nothing about that ^. And it **must not** know that.
It's only `ChePluginService` who has this knowledge about plug-ins, icons and so on.

Comment (azatsarynnyy) [reply extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts:75]: Let's imagine, in some other place in the code, I call `workspaceService.getWorkspaceSettings()`
in order to see if `privateURI` is absent. Despite the presence of `publicURI`.
How I can know that? There's no chance.
That's why I'm proposing to return the `WorkspaceSettings` as it is. And keep such checking only in the place where it's actually important - like in `ChePluginService`.
